### PR TITLE
Add int8 resnet50 model

### PR DESCRIPTION
**Pull request by mengniwang95**
# ResNet

## Use cases
ResNet models perform image classification - they take images as input and classify the major object in the image into a set of pre-defined classes. They are trained on ImageNet dataset which contains images from 1000 classes. ResNet models provide very high accuracies with affordable model sizes. They are ideal for cases when high accuracy of classification is required.

## Description
Deeper neural networks are more difficult to train. Residual learning framework ease the training of networks that are substantially deeper. The research explicitly reformulate the layers as learning residual functions with reference to the layer inputs, instead of learning unreferenced functions. It also provide comprehensive empirical evidence showing that these residual networks are easier to optimize, and can gain accuracy from considerably increased depth. On the ImageNet dataset the residual networks are evaluated with a depth of up to 152 layers - 8X deeper than VGG nets but still having lower complexity.

ONNX ResNet50-v1 ==> Quantized ONNX ResNet50-v1

## Model

The model below is ResNet50-v1. ResNet50_int8 is obtained by quantizing ResNet50_fp32 model. Note the latency depends on the test hardware and the displayed results are tested by Intel® Xeon® Platinum 8280 Processor. 

Compared with the fp32 model, we get an accuracy drop of 0.2% and a performance improvement of 1.9X after quantization.

|Model        |Download  |Download (with sample test data)| ONNX version |Opset version|Top-1 accuracy (%)|latency (ms)|
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|ResNet50_fp32|[97.8 MB](model/resnet50-v1-12.onnx)|[92.0 MB](model/resnet50-v1-12.tar.gz)|1.7.0|12|74.97|8.32|
|ResNet50_int8|[24.6 MB](model/resnet50-v1-12-int8.onnx)|[22.3 MB](model/resnet50-v1-12-int8.tar.gz)|1.7.0|12|74.77|4.30|

## Inference
We use onnxruntime to perform inference. View the notebook [onnxrt_inference](../../onnxrt_inference.ipynb) to understand how to use above models for doing inference.

### Input
The fp32 and int8 models expect input images normalized in the same way, i.e. mini-batches of 3-channel RGB images of shape (N x 3 x H x W), where N is the batch size expected to be 1, and H and W are expected to be 224.

### Preprocessing
```python
def preprocess(img):
    img = img / 255.
    img = cv2.resize(img, (256, 256))
    h, w = img.shape[0], img.shape[1]
    y0 = (h - 224) // 2
    x0 = (w - 224) // 2
    img = img[y0 : y0+224, x0 : x0+224, :]
    img = (img - [0.485, 0.456, 0.406]) / [0.229, 0.224, 0.225]
    img = np.transpose(img, axes=[2, 0, 1])
    img = img.astype(np.float32)
    img = np.expand_dims(img, axis=0)
    return img
```

### Output
The model outputs image scores for each of the [1000 classes of ImageNet](../../synset.txt).

### Postprocessing
The post-processing sort the predictions to report the most probable classes.

```python
preds = np.squeeze(preds)
a = np.argsort(preds)[::-1]
print('class=%s ; probability=%f' %(labels[a[0]],preds[a[0]]))
```

### Dataset
Dataset used for validation: [ImageNet (ILSVRC2012)](http://www.image-net.org/challenges/LSVRC/2012/). Check [imagenet_prep](../../imagenet_prep.md) for guidelines on preparing the dataset.

## Quantization
We use [Intel® Low Precision Optimization Tool (LPOT)](https://github.com/intel/lpot) with onnxruntime backend to perform quantization. View the [instructions](https://github.com/intel/lpot/tree/master/examples/onnxrt/onnx_model_zoo/resnet50/README.md) to understand how to use LPOT for quantization.

### Environment
onnx: 1.7.0 
onnxruntime: 1.6.0+

### Prepare model
```shell
wget https://github.com/onnx/models/raw/master/vision/classification/resnet/model_quantized/model/resnet50-v1-12.onnx
```

### Model quantize
Make sure to specify the appropriate dataset path in the configuration file.
```bash
bash run_tuning.sh --input_model=path/to/model \  # model path as *.onnx
                   --config=resnet50_v1_5.yaml \
                   --output_model=path/to/save
```

### References
* **ResNetv1**
[Deep residual learning for image recognition](https://arxiv.org/abs/1512.03385)
 He, Kaiming, Xiangyu Zhang, Shaoqing Ren, and Jian Sun. In Proceedings of the IEEE conference on computer vision and pattern recognition, pp. 770-778. 2016.

* Intel® Low Precision Optimization Tool (LPOT): https://github.com/intel/lpot

## Contributors
* [mengniwang95](https://github.com/mengniwang95) (Intel)
* [airMeng](https://github.com/airMeng) (Intel)
* [ftian1](https://github.com/ftian1) (Intel)
* [hshen14](https://github.com/hshen14) (Intel)

## License
Apache 2.0

### CODE REVIEW

**Comment by wenbingl** [comment]
@mengniwang95 , it looks there are some failure on running onnx.checker for the quantized model, can you take a look?

**Comment by mengniwang95** [comment]
> @mengniwang95 , it looks there are some failure on running onnx.checker for the quantized model, can you take a look?

sure, I will check it asap

**Comment by wenbingl** [comment]
@mengniwang95, I guess the output mismatch for resnet_int8 is expected, can you double confirm that. 

@jcwchen, any idea how to specify a test data for the quantized model since which output always has some noticeable gap from the float32.

**Comment by jcwchen** [comment]
> @mengniwang95, I guess the output mismatch for resnet_int8 is expected, can you double confirm that.
> 
> @jcwchen, any idea how to specify a test data for the quantized model since which output always has some noticeable gap from the float32 version.

Could you please elaborate more about the use case? Why not just update the output data in the test directory?

**Comment by wenbingl** [comment]
> > @mengniwang95, I guess the output mismatch for resnet_int8 is expected, can you double confirm that.
> > @jcwchen, any idea how to specify a test data for the quantized model since which output always has some noticeable gap from the float32 version.
> 
> Could you please elaborate more about the use case? Why not just update the output data in the test directory?

The problem is there is a floating point version model in the same directory.

**Comment by jcwchen** [comment]
Still, I think the new output.pb under test_data_set_0 can be updated by the floating point version model. You should be able to use the function `create_test_dir` [here](https://github.com/microsoft/onnxruntime/blob/067759b387fd1f789d0dc4755b9d7fd3f7063bf9/tools/python/ort_test_dir_utils.py#L66) to do so. Please let me know if I misunderstand. Thanks!

**Comment by mengniwang95** [comment]
I use https://github.com/microsoft/onnxruntime/blob/067759b387fd1f789d0dc4755b9d7fd3f7063bf9/tools/python/ort_test_dir_utils.py#L66 to create the input and output data of fp32&int8 models, and I can pass the test on my local machine. I will double-check the output data. Thx

**Comment by mengniwang95** [comment]
@jcwchen I am a little confused, I prepared a conda env with same python version and same dependences, and set the model_list of https://github.com/onnx/models/blob/bb0d4cf3d4e2a5f7376c13a08d337e86296edbe8/workflow_scripts/test_models.py#L31 as '/home/mengniwa/onnx_model_zoo/models/vision/classification/resnet/model_quantized/model/resnet50-v1-12-int8.onnx' to check model. I found the actual output of int8 model with test_data_2 is the same as the expected data. 
This is the output of script:
==============Testing resnet50-v1-12-int8.onnx==============
/home/mengniwa/onnx_model_zoo/models/vision/classification/resnet/model_quantized/model/resnet50-v1-12-int8.tar.gz
LFS pull completed with return code= 1
Running tests in /home/mengniwa/onnx_model_zoo/models/workflow_scripts/ci_test_tar_dir/resnet50-v1-12-int8 for /home/mengniwa/onnx_model_zoo/models/workflow_scripts/ci_test_tar_dir/resnet50-v1-12-int8/resnet50-v1-12-int8.onnx
/home/mengniwa/onnx_model_zoo/models/workflow_scripts/ci_test_tar_dir/resnet50-v1-12-int8/test_data_set_0
PASS
/home/mengniwa/onnx_model_zoo/models/workflow_scripts/ci_test_tar_dir/resnet50-v1-12-int8/test_data_set_1
PASS
/home/mengniwa/onnx_model_zoo/models/workflow_scripts/ci_test_tar_dir/resnet50-v1-12-int8/test_data_set_2
PASS
[PASS] resnet50-v1-12-int8.tar.gz is checked by onnxruntime.
LFS pull completed with return code= 1
[PASS] resnet50-v1-12-int8.onnx is checked by onnx.
1 models have been checked.

If there is any wrong step? Thanks

**Comment by jcwchen** [comment]
Hi @mengniwang95,
Which ORT version did you use? I tried ORT 1.8.0 and 1.8.1 in my local, `ort_test_dir_utils.run_test_dir('resnet50-v1-12-int8')` also showed output mismatch.

**Comment by mengniwang95** [comment]
> Hi @mengniwang95,
> Which ORT version did you use? I tried ORT 1.8.0 and 1.8.1 in my local, `ort_test_dir_utils.run_test_dir('resnet50-v1-12-int8')` also showed output mismatch.

I use ORT 1.8.0

**Comment by jcwchen** [comment]
On my Mac, I can repro the same error as Windows-CI and Linux-CI with the same output values. And the difference looks significant... I am not familiar with int8 kernel in ORT... @wenbingl Do you know whether it is possible that ORT inference with int8 produces different values in different environments?

@mengniwang95 Could you please provide more details about your environment (like OS, numpy, ONNX version)? Thank you.

**Comment by mengniwang95** [comment]
> On my Mac, I can repro the same error as Windows-CI and Linux-CI with the same output values. And the difference looks significant... I am not familiar with int8 kernel in ORT... @wenbingl Do you know whether it is possible that ORT inference with int8 produces different values in different environments?
> 
> @mengniwang95 Could you please provide more details about your environment (like OS, numpy, ONNX version)? Thank you.

OS: CentOS Linux release 7.6.1810

Package           Version
----------------- -------------------
certifi           2021.5.30
chardet           4.0.0
flatbuffers       2.0
idna              2.10
numpy             1.21.0
onnx              1.7.0
onnxruntime       1.8.0
pip               21.1.3
protobuf          3.17.3
requests          2.25.1
setuptools        49.6.0.post20210108
six               1.16.0
typing-extensions 3.10.0.0
urllib3           1.26.6
wheel             0.36.2

**Comment by jcwchen** [comment]
@mengniwang95  Could you please try this model in your environment (`ort_test_dir_utils.run_test_dir('resnet50-v1-12-int8')`) and see whether there is any error? Thanks.
[resnet50-v1-12-int8.tar.gz](https://github.com/onnx/models/files/6816885/resnet50-v1-12-int8.tar.gz)

**Comment by mengniwang95** [comment]
> @mengniwang95 Could you please try this model in your environment (`ort_test_dir_utils.run_test_dir('resnet50-v1-12-int8')`) and see whether there is any error? Thanks.
> [resnet50-v1-12-int8.tar.gz](https://github.com/onnx/models/files/6816885/resnet50-v1-12-int8.tar.gz)

@jcwchen I try this model and produce mismatch error. Thanks

**Comment by mengniwang95** [comment]
@jcwchen My output data is got with VNNI instructions, if this instruction is available on the test machine? If not, pls offer me the machine info and I can provide consistent output data. Thanks

**Comment by jcwchen** [comment]
> @jcwchen My output data is got with VNNI instructions, if this instruction is available on the test machine? If not, pls offer me the machine info and I can provide consistent output data. Thanks

I don't think the GitHub Action machines have VNNI instructions. I would say they are using general CPU machines. Do you have any handy non-VNNI machine to produce/test the output data? Then we can verify whether VNNI causes this output mismatch. If you don't have such a machine, I can provide the test data from my side (If you are comfortable with it). Thank you!

**Comment by mengniwang95** [comment]
@jcwchen I have updated the data generated by non VNNI machine. But I have one question, maybe you can add some check codes in model check script to confirm the cpu info ? Because the cpu type will affect the test result... Thanks

@wenbingl Currently I add the int8 model readme under /resnet/model_quantized/, if users do not enter this path they won't see this introduction. I am not sure whether there is a better path to put the introduction & models ? Thanks

**Comment by jcwchen** [comment]
> @jcwchen I have updated the data generated by non VNNI machine.

Thanks for the update! They now all pass.

> But I have one question, maybe you can add some check codes in model check script to confirm the cpu info ? Because the cpu type will affect the test result... Thanks

Good point, but I didn't find a proper way (standard library) to detect VNNI in Python-level... We will keep this in mind to prevent similar future uploading issues. Thank you for the contribution!

**Comment by wenbingl** [comment]
> @jcwchen I have updated the data generated by non VNNI machine. But I have one question, maybe you can add some check codes in model check script to confirm the cpu info ? Because the cpu type will affect the test result... Thanks
> 
> @wenbingl Currently I add the int8 model readme under /resnet/model_quantized/, if users do not enter this path they won't see this introduction. I am not sure whether there is a better path to put the introduction & models ? Thanks

Thanks for the contribution, @mengniwang95. 
To add more exposure for the quantized model, you can add some words and a link to int8 model  in https://github.com/onnx/models/tree/master/vision/classification/resnet/README.md. I will merge this PR first and you can post another PR to add the link.

**Review by wenbingl — APPROVED** [review]

**Comment by hshen14** [comment]
@wenbingl Hi Wenbing, could you please help revert the merged PR asap? We need to wait for some process and re-submit the PR after that. Sorry for the inconvenience and thanks for your understanding.
